AppController/djinn.rb

```diff
@@ -3550,60 +3447,86 @@ def start_api_services()
           @state = "Couldn't start Zookeeper."
           HelperFunctions.log_and_crash(@state, WAIT_TO_CRASH)
         end
-        start_backup_service()
       end
-
+      write_zookeeper_locations()
       ZKInterface.init(my_node, @nodes)
+      Djinn.log_info("Done configuring zookeeper.")
     }
 
-    if my_node.is_memcache?
+    if my_node.is_db_master? or my_node.is_db_slave?
       threads << Thread.new {
-        start_memcache()
+        Djinn.log_info("Starting database services.")
+        if my_node.is_db_master?
+          start_db_master(@options['clear_datastore'].downcase == "true")
+          prime_database
+        else
+          start_db_slave(@options['clear_datastore'].downcase == "true")
+        end
+        # Let's make sure cassandra is up.
+        HelperFunctions.sleep_until_port_is_open(@my_private_ip,
+          THRIFT_PORT)
+
+        # Always colocate the Datastore Server and UserAppServer (soap_server).
+        @state = "Starting up SOAP Server and Datastore Server"
+        start_datastore_server()
+
+        # Start the UserAppServer and wait till it's ready.
+        start_soap_server()
+        Djinn.log_info("Done starting database services.")
       }
     end
 
-    if my_node.is_db_master?
+    # We now wait for the essential services to go up.
+    Djinn.log_info("Waiting for DB services ... ")
+    threads.each { |t| t.join() }
+    @done_initializing = true
+
+    # All nodes waits for the UserAppServer now.
+    configure_uaserver_nginx()
+    HelperFunctions.sleep_until_port_is_open(@my_private_ip,
+      UserAppClient::SERVER_PORT, USE_SSL)
+    uac = UserAppClient.new(@my_private_ip, @@secret)
+    begin
+      app_list = uac.get_all_apps()
+    rescue FailedNodeException
+      Djinn.log_debug("UserAppServer not ready yet: retrying.")
+      retry
+    end
+    @userappserver_ip = my_node.private_ip
+    Djinn.log_info("UserAppServer is ready.")
+
+    # We can not start all the other services that may depends on the
```

Comment:
@menivaitsi is this a question on the grammar? or the logic? :)

if it's grammar, please let me know what's best.

If it's logic, yes some services depends on the UAServer to be fully operational. With the way we do it with this patch. checking the ports it's not enough to see if it's up, so we need to do a full check to confirm that the service is operational. If we move ahead, we risked the tools to do operations too soon,